plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/internal/refresh/layout/SiriusCanonicalLayoutHandler.java

```diff
@@ -195,35 +204,239 @@ private static List<IAdaptable> getAdapters(Set<View> createdViewsToLayout) {
+    /**
+     * Gets BorderNodes of an adapted view.
+     * 
+     * @param viewAdapter
+     *            the adapted view
+     * @return the set of BorderNodes of the adapted view.
+     */
+    private static Set<DNode> getBorderNodes(IAdaptable viewAdapter) {
+        Set<DNode> borderNodes = new HashSet<>();
+        View createdView = viewAdapter.getAdapter(View.class);
+        if (createdView != null && createdView.getElement() instanceof AbstractDNode dnode) {
+            borderNodes.addAll(dnode.getOwnedBorderedNodes());
+        }
+        return borderNodes;
+    }
+
+    /**
+     * Gets BorderNodes of an EditPart.
+     * 
+     * @param elementEditPart
+     *            the EditPart
+     * @return the set of BorderNodes of the EditPart.
+     */
+    private static Set<DNode> getBorderNodes(IDiagramElementEditPart elementEditPart) {
+        Set<DNode> borderNodes = new HashSet<>();
+        if (elementEditPart.resolveDiagramElement() instanceof AbstractDNode dnode) {
+            borderNodes.addAll(dnode.getOwnedBorderedNodes());
+        }
+        return borderNodes;
+    }
+
+    /**
+     * Check if an EditPart has BorderNodes.
+     * 
+     * @param elementEditPart
+     *            the EditPart
+     * @return {@code true} if the EditPart has BorderNodes; {@code false} otherwise.
+     */
+    private static boolean hasBorderNodes(IDiagramElementEditPart elementEditPart) {
+        return !getBorderNodes(elementEditPart).isEmpty();
+    }
+
+    /**
+     * Check if an adapted view has BorderNodes.
+     * 
+     * @param viewAdapter
+     *            the adapted view
+     * @return {@code true} if the adapted view has BorderNodes; {@code false} otherwise.
+     */
+    private static boolean hasBorderNodes(IAdaptable viewAdapter) {
+        return !getBorderNodes(viewAdapter).isEmpty();
+    }
+
+    private static Map<IGraphicalEditPart, List<IAdaptable>> getBorderedNodesFromCompartment(AbstractDNodeContainerCompartmentEditPart compartmentEP, List<IAdaptable> viewAdapters) {
+        Map<IGraphicalEditPart, List<IAdaptable>> borderedNodesToLayoutMap = new LinkedHashMap<IGraphicalEditPart, List<IAdaptable>>();
+        for (IAdaptable viewAdapter : viewAdapters) {
+            if (hasBorderNodes(viewAdapter)) {
+                View createdView = viewAdapter.getAdapter(View.class);
+                if (createdView != null && createdView.getElement() instanceof AbstractDNode dnode) {
+                    List<IAdaptable> existingList = borderedNodesToLayoutMap.get(compartmentEP);
+                    if (existingList == null) {
+                        existingList = new ArrayList<>();
+                    }
+                    existingList.add(viewAdapter);
+                    borderedNodesToLayoutMap.put(compartmentEP, existingList);
+                }
+            }
+        }
+        return borderedNodesToLayoutMap;
+    }
```

Comment:
Perhaps I haven't been specific enough about the things I find strange about this function: I don't understand why the map intervenes in the loop (L266, L271) when in the end there's only one element in this map since the key is compartmentEP and this variable doesn't change throughout the loop. Wouldn't code like this be easier to read?

```java
List<IAdaptable> existingList = new ArrayList<>();
for (IAdaptable viewAdapter : viewAdapters) {
    if (hasBorderNodes(viewAdapter)) {
        View createdView = viewAdapter.getAdapter(View.class);
        if (createdView != null && createdView.getElement() instanceof AbstractDNode dnode) {
            existingList.add(viewAdapter);
        }
    }
}

Map<IGraphicalEditPart, List<IAdaptable>> borderedNodesToLayoutMap = new LinkedHashMap<IGraphicalEditPart, List<IAdaptable>>();
if (existingList.isEmpty()) {
    borderedNodesToLayoutMap.put(compartmentEP, existingList);
}
return borderedNodesToLayoutMap;
```

Or even functional with a simple filter?

```java
List<IAdaptable> existingList = viewAdapters.stream().filter(viewAdapter -> {
    View createdView = viewAdapter.getAdapter(View.class);
    return hasBorderNodes(viewAdapter) && createdView != null && createdView.getElement() instanceof AbstractDNode;
}).collect(Collectors.toList());

Map<IGraphicalEditPart, List<IAdaptable>> borderedNodesToLayoutMap = new LinkedHashMap<IGraphicalEditPart, List<IAdaptable>>();
if (existingList.isEmpty()) {
    borderedNodesToLayoutMap.put(compartmentEP, existingList);
}
return borderedNodesToLayoutMap;
```